tutorials/migrating/upgrading_to_godot_4.rst

```diff
@@ -509,13 +509,14 @@ environment effect and its visual knobs remain within the Environment resource.
 Updating shaders
 ^^^^^^^^^^^^^^^^
 
-There have been some changes to shaders that aren't covered by the upgrade tool.
+There have been some changes to shaders that aren't covered by the upgrade tool. 
+You will need to make some manual changes, especially if your shader is advanced.
```

Comment:
... especially if your shader _performs advanced operations_ ?

Reply:
"Advanced operations" isn't *quite* right, either. What I want here is a disclaimer that you still need to make manual changes, and that the list here can't catch literally every change.

Things that should be covered by "advanced" or its replacement, without explicitly listed them out:
- complex `light()` functions that make assumptions about defaults or the standard shader lighting function
- assumptions about the specifics of coordinate spaces, like the NDC change or the internals of the transform matrixes. 
- using clip space directly (all the cases in the reverse-z blogpost that don't have example snippets)

Just using a transform matrix to convert coordinates is not an "advanced operation", but taking the matrix apart and manually reassembling it for some other purpose is.

I should probably just add an explicit note that this list is *not* exhaustive, and your shader may not work even if all the changes are made.

Reply:
I think you can simplify and say something like _especially if your shader uses a custom `light()` function or includes coordinate space transformations._

Advanced users will be flagged about the risks, while beginner users will know that those are advanced things that they are not using.

Reply:
Yeah I get it. I guess there is no sweet spot as it's hard to qualify what may require some rewrite and what may not.
It's just that "advanced shader" doesn't speak much to me (the shader itself is not advanced, but more the actions it performs, or the use you make of the shading functionalities). But I'm not a native speaker so others may feel differently.

Reply:
I ended up using the specific cases that clayjohn mentioned, rather than trying to generalize too much.

I also added another paragraph at the end which solves what I saw as the actual problem.